File: pipeline/dp_cpp/src/bellman_locked.cpp
#include "bellman_locked.hpp"
#include <algorithm>
#include <cmath>
#include <limits>
#include <numeric>
#include <set>
#include <stdexcept>
#include <unordered_set>

static constexpr double SOG_TOL           = 1e-6;
static constexpr double LOCK_KEY_PRECISION = 1e6;

static double round_lock(double sog) {
    return std::round(sog * LOCK_KEY_PRECISION) / LOCK_KEY_PRECISION;
}

BellmanSolverLocked::BellmanSolverLocked(const std::vector<Node>& nodes,
                                           const std::vector<AtomicEdge>& edges,
                                           const std::set<double>& v_line_times)
    : edges_(edges) {
    std::unordered_set<int64_t> v_t_set;
    v_t_set.insert(0LL);
    for (double t : v_line_times)
        v_t_set.insert(static_cast<int64_t>(std::llround(t * 1e9)));

    for (auto& n : nodes) {
        TDKey k = make_td_key(n.time_h, n.distance_nm);
        auto it = key_to_id_.find(k);
        if (it == key_to_id_.end()) {
            int id = (int)coords_.size();
            key_to_id_[k] = id;
            coords_.push_back({n.time_h, n.distance_nm});
            is_source_.push_back(n.is_source);
            is_sink_.push_back(n.is_sink);
            is_v_line_.push_back(v_t_set.count(k.t9) > 0);
        } else {
            int id = it->second;
            if (n.is_source) is_source_[id] = true;
            if (n.is_sink)   is_sink_[id]   = true;
        }
    }

    int N = (int)coords_.size();
    outgoing_.resize(N);
    for (int ei = 0; ei < (int)edges_.size(); ++ei) {
        auto it = key_to_id_.find(make_td_key(edges_[ei].src_t, edges_[ei].src_d));
        if (it == key_to_id_.end()) { ++unknown_edges_; continue; }
        outgoing_[it->second].push_back(ei);
    }

    state_.resize(N);
    source_id_ = -1;
    for (int i = 0; i < N; ++i)
        if (is_source_[i]) { source_id_ = i; break; }
    if (source_id_ < 0) {
        auto it = key_to_id_.find(make_td_key(0.0, 0.0));
        if (it == key_to_id_.end()) throw std::runtime_error("No source node at (0,0)");
        source_id_ = it->second;
    }
    // Source starts with lock = UNLOCKED, cost = 0, no parent
    LockEntry src_entry;
    src_entry.cost = 0.0;
    state_[source_id_][UNLOCKED] = src_entry;

    topo_order_.resize(N);
    std::iota(topo_order_.begin(), topo_order_.end(), 0);
    std::sort(topo_order_.begin(), topo_order_.end(),
              [&](int a, int b) { return coords_[a] < coords_[b]; });
}

void BellmanSolverLocked::solve() {
    for (int src_id : topo_order_) {
        if (state_[src_id].empty()) continue;

        // Snapshot so dst writes don't invalidate iteration
        LockMap snapshot = state_[src_id];

        for (auto& [lock_state, src_entry] : snapshot) {
            double base_cost = src_entry.cost;

            for (int ei : outgoing_[src_id]) {
                const auto& e = edges_[ei];
                if (std::isnan(e.fuel_mt)) { ++nan_edges_skipped_; continue; }

                // Lock admissibility
                double new_lock_in_block;
                if (lock_state == UNLOCKED) {
                    new_lock_in_block = round_lock(e.target_sog);
                } else {
                    if (std::abs(e.target_sog - lock_state) > SOG_TOL) continue;
                    new_lock_in_block = lock_state;
                }

                auto dit = key_to_id_.find(make_td_key(e.dst_t, e.dst_d));
                if (dit == key_to_id_.end()) continue;
                int dst_id = dit->second;

                double new_lock = is_v_line_[dst_id] ? UNLOCKED : new_lock_in_block;
                double new_cost = base_cost + e.fuel_mt;

                auto& dst_map = state_[dst_id];
                auto mit = dst_map.find(new_lock);
                if (mit == dst_map.end() || new_cost < mit->second.cost) {
                    LockEntry entry;
                    entry.cost      = new_cost;
                    entry.edge_idx  = ei;
                    entry.prev_node = src_id;
                    entry.prev_lock = lock_state;
                    dst_map[new_lock] = entry;
                }
            }
        }
    }
}

int BellmanSolverLocked::best_sink(double eta_h) {
    int best_id   = -1;
    double best_c = std::numeric_limits<double>::infinity();
    for (int i = 0; i < (int)coords_.size(); ++i) {
        if (!is_sink_[i]) continue;
        if (coords_[i].first > eta_h + 1e-6) continue;
        auto it = state_[i].find(UNLOCKED);
        if (it == state_[i].end()) continue;
        if (it->second.cost < best_c) {
            best_c  = it->second.cost;
            best_id = i;
        }
    }
    if (best_id < 0) throw std::runtime_error("No sink reachable within ETA under Luo lock.");
    return best_id;
}

std::vector<int> BellmanSolverLocked::backtrack(int sink_id) {
    std::vector<int> path;
    int    cur_id   = sink_id;
    double cur_lock = UNLOCKED;
    for (;;) {
        auto mit = state_[cur_id].find(cur_lock);
        if (mit == state_[cur_id].end() || !mit->second.has_parent()) break;
        const auto& e = mit->second;
        path.push_back(e.edge_idx);
        cur_id   = e.prev_node;
        cur_lock = e.prev_lock;
        if (cur_id == source_id_ && cur_lock == UNLOCKED) break;
    }
    std::reverse(path.begin(), path.end());
    return path;
}

LuoResult BellmanSolverLocked::result(double eta_h) {
    int sink_id      = best_sink(eta_h);
    auto schedule    = backtrack(sink_id);
    double sink_cost = state_[sink_id].at(UNLOCKED).cost;

    int states_reached = 0;
    std::set<double> distinct_locks;
    for (auto& lmap : state_) {
        states_reached += (int)lmap.size();
        for (auto& [lock, _v] : lmap)
            if (lock != UNLOCKED) distinct_locks.insert(lock);
    }

    return {sink_cost, coords_[sink_id].first,
            schedule, coords_[sink_id],
            nan_edges_skipped_, states_reached,
            (int)distinct_locks.size()};
}

File: pipeline/dp_rebuild/bellman_locked.py
"""
Bellman DP solver with per-block SOG-lock state for Luo 2024 mode.

State = (canonical_node_id, locked_target_sog | None):
  * V-line node:     lock = None (forecast block boundary; new SOG can be picked)
  * H-line node mid-block: lock = S (the target_sog chosen at the block-start V-line)

Transitions:
  * Outgoing from (V-line src, None): any atomic edge admissible;
                                       sets new lock = edge.target_sog
                                       (or None again if dst is itself a V-line)
  * Outgoing from (H-line src, S):    only edges with edge.target_sog == S
                                       admissible; lock stays S until next V-line

Free DP and Luo DP run on the SAME atomic-edge graph. Difference is purely
in this Bellman state augmentation.

Spec reference: docs/meeting_prep_2026_05_11.md §2.1.3 (Bellman mode summary).
"""

from __future__ import annotations

from dataclasses import dataclass
from math import inf, isnan
from typing import Dict, List, Optional, Set, Tuple

from atomic_edges import AtomicEdge
from nodes import Node


_KEY_PRECISION = 9
_SOG_TOLERANCE = 1e-6


def _key(t: float, d: float) -> Tuple[float, float]:
    return (round(t, _KEY_PRECISION), round(d, _KEY_PRECISION))


@dataclass
class LuoBellmanResult:
    total_fuel_mt: float
    voyage_time_h: float
    schedule: List[AtomicEdge]
    sink_node: Tuple[float, float]
    nan_edges_skipped: int
    states_reached: int
    distinct_locks_used: int


class BellmanSolverLocked:
    """Forward Bellman over (canonical_node, locked_sog) states."""

    def __init__(
        self,
        nodes: List[Node],
        edges: List[AtomicEdge],
        v_line_times: Set[float],
    ):
        # ---- canonicalise nodes ----
        seen: Dict[Tuple[float, float], int] = {}
        coords: List[Tuple[float, float]] = []
        is_source: List[bool] = []
        is_sink: List[bool] = []
        is_v_line: List[bool] = []

        v_set_rounded = {round(t, _KEY_PRECISION) for t in v_line_times}
        v_set_rounded.add(round(0.0, _KEY_PRECISION))  # source sits on implicit V-line

        for n in nodes:
            k = _key(n.time_h, n.distance_nm)
            if k not in seen:
                seen[k] = len(coords)
                coords.append(k)
                is_source.append(n.is_source)
                is_sink.append(n.is_sink)
                is_v_line.append(k[0] in v_set_rounded)
            else:
                idx = seen[k]
                if n.is_source:
                    is_source[idx] = True
                if n.is_sink:
                    is_sink[idx] = True
                # is_v_line is by coordinate, not flag — already set correctly

        n_canonical = len(coords)
        self._coords = coords
        self._key_to_id = seen
        self._is_source = is_source
        self._is_sink = is_sink
        self._is_v_line = is_v_line

        # ---- index outgoing edges ----
        self._outgoing: List[List[AtomicEdge]] = [[] for _ in range(n_canonical)]
        self._unknown_edges = 0
        for e in edges:
            src_id = seen.get(_key(e.src_t, e.src_d))
            if src_id is None:
                self._unknown_edges += 1
                continue
            self._outgoing[src_id].append(e)

        # ---- source ----
        explicit_sources = [i for i, s in enumerate(is_source) if s]
        if explicit_sources:
            self._source_id = explicit_sources[0]
        else:
            zid = seen.get(_key(0.0, 0.0))
            if zid is None:
                raise ValueError("No source node at (0, 0)")
            self._source_id = zid

        # ---- topological order ----
        self._topo_order = sorted(range(n_canonical), key=lambda i: coords[i])

        # ---- DP state: per-node dict {lock -> (cost, parent_record)} ----
        # parent_record = (edge, prev_canonical_id, prev_lock) or None for source
        self._state: List[Dict[Optional[float], Tuple[float, Optional[Tuple[AtomicEdge, int, Optional[float]]]]]] = [
            {} for _ in range(n_canonical)
        ]
        self._state[self._source_id][None] = (0.0, None)

        self._nan_edges_skipped = 0

    # ------------------------------------------------------------------

    def solve(self) -> None:
        """Forward sweep with SOG-lock state augmentation."""
        is_v_line = self._is_v_line

        for src_id in self._topo_order:
            states = self._state[src_id]
            if not states:
                continue

            # snapshot — we mutate `self._state[dst_id]` below; src state never
            # changes during its own iteration, so a list of (lock, cost) is safe
            src_states = list(states.items())

            for lock_state, (base_cost, _parent) in src_states:
                for e in self._outgoing[src_id]:
                    if isnan(e.fuel_mt):
                        self._nan_edges_skipped += 1
                        continue

                    # ---- lock admissibility ----
                    if lock_state is None:
                        new_lock_in_block = round(e.target_sog, 6)
                    else:
                        if abs(e.target_sog - lock_state) > _SOG_TOLERANCE:
                            continue
                        new_lock_in_block = lock_state

                    # ---- resolve dst ----
                    dst_id = self._key_to_id.get(_key(e.dst_t, e.dst_d))
                    if dst_id is None:
                        continue

                    # Reaching a V-line node releases the lock
                    new_lock = None if is_v_line[dst_id] else new_lock_in_block

                    new_cost = base_cost + e.fuel_mt
                    cur = self._state[dst_id].get(new_lock)
                    if cur is None or new_cost < cur[0]:
                        self._state[dst_id][new_lock] = (
                            new_cost,
                            (e, src_id, lock_state),
                        )

    # ------------------------------------------------------------------

    def best_sink(self, eta_h: float) -> int:
        """Argmin sink reachable with lock=None at t ≤ ETA."""
        best_id = -1
        best_cost = inf
        for i, sink in enumerate(self._is_sink):
            if not sink:
                continue
            if self._coords[i][0] > eta_h + 1e-6:
                continue
            entry = self._state[i].get(None)
            if entry is None:
                continue
            if entry[0] < best_cost:
                best_cost = entry[0]
                best_id = i
        if best_id < 0:
            raise ValueError("No sink reachable within ETA under Luo lock.")
        return best_id

    # ------------------------------------------------------------------

    def backtrack(self, sink_id: int) -> List[AtomicEdge]:
        """Walk parent pointers from (sink_id, None) back to source."""
        path: List[AtomicEdge] = []
        cur_id = sink_id
        cur_lock: Optional[float] = None
        while True:
            entry = self._state[cur_id].get(cur_lock)
            if entry is None or entry[1] is None:
                break
            edge, prev_id, prev_lock = entry[1]
            path.append(edge)
            cur_id = prev_id
            cur_lock = prev_lock
            if cur_id == self._source_id and cur_lock is None:
                break
        path.reverse()
        return path

    # ------------------------------------------------------------------

    def result(self, eta_h: float) -> LuoBellmanResult:
        sink_id = self.best_sink(eta_h)
        sink_t, sink_d = self._coords[sink_id]
        schedule = self.backtrack(sink_id)
        states_reached = sum(len(s) for s in self._state)
        distinct_locks = len({lock for s in self._state for lock in s if lock is not None})
        sink_cost = self._state[sink_id][None][0]
        return LuoBellmanResult(
            total_fuel_mt=float(sink_cost),
            voyage_time_h=float(sink_t),
            schedule=schedule,
            sink_node=(float(sink_t), float(sink_d)),
            nan_edges_skipped=self._nan_edges_skipped,
            states_reached=states_reached,
            distinct_locks_used=distinct_locks,
        )

    @property
    def num_unknown_edges(self) -> int:
        return self._unknown_edges
